internal/controllers/resourceslice/slice.go

```diff
@@ -0,0 +1,197 @@
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(fmt.Errorf("getting composition: %w", err))
+	}
+	logger = logger.WithValues("compositionGeneration", comp.Generation, "compositionName", comp.Name, "compositionNamespace", comp.Namespace, "synthesisID", comp.Status.GetCurrentSynthesisUUID())
+	ctx = logr.NewContext(ctx, logger)
+
+	if comp.Status.CurrentSynthesis == nil {
+		return ctrl.Result{}, nil
+	}
+
+	snapshot := statusSnapshot{Reconciled: true, Ready: true}
+
+	for _, ref := range comp.Status.CurrentSynthesis.ResourceSlices {
+		slice := &apiv1.ResourceSlice{}
+		slice.Name = ref.Name
+		slice.Namespace = comp.Namespace
+		err := s.client.Get(ctx, client.ObjectKeyFromObject(slice), slice)
+		if errors.IsNotFound(err) {
+			if comp.DeletionTimestamp != nil {
+				logger.V(1).Info("resource slice is missing, ignoring because composition is being deleted", "resourceSliceName", ref.Name)
+				continue
+			}
+			return s.handleMissingSlice(ctx, comp, ref.Name)
+		}
+		if err != nil {
+			return ctrl.Result{}, client.IgnoreNotFound(fmt.Errorf("getting resource slice: %w", err))
+		}
+
+		// Handle a case where the reconciliation controller hasn't updated the slice's status yet
+		if len(slice.Status.Resources) == 0 && len(slice.Spec.Resources) > 0 {
+			snapshot.Ready = false
+			snapshot.Reconciled = false
+			break // no need to check the other slices
+		}
+
+		// Collect the state of every resource
+		for _, state := range slice.Status.Resources {
+			state := state
+			if resourceNotReconciled(comp, &state) {
+				snapshot.Reconciled = false
+			}
+			if state.Ready == nil {
+				snapshot.Ready = false
+			}
+			if state.Ready != nil && (snapshot.ReadyTime == nil || snapshot.ReadyTime.Before(state.Ready)) {
+				snapshot.ReadyTime = state.Ready
+			}
+		}
+	}
+
+	// Aggregate the status of all slices into the composition
+	if !processCompositionTransition(ctx, comp, snapshot) {
+		return ctrl.Result{}, nil
+	}
+	err = s.client.Status().Update(ctx, comp)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("updating composition '%s' status: %w", comp.Name, err)
+
+	}
+	logger.V(1).Info("aggregated resource status into composition", "reconciled", snapshot.Reconciled, "ready", snapshot.Ready)
+
+	return ctrl.Result{}, nil
+}
+
+func (s *sliceController) handleMissingSlice(ctx context.Context, comp *apiv1.Composition, sliceName string) (ctrl.Result, error) {
+	logger := logr.FromContextOrDiscard(ctx)
+
+	// We can't do anything about missing resource slices if synthesis is already in-flight or it isn't safe to resynthesize
+	if comp.ShouldIgnoreSideEffects() || comp.Status.InFlightSynthesis != nil || comp.ShouldForceResynthesis() {
+		return ctrl.Result{}, nil
+	}
+
+	// It's possible that newly created slices haven't hit the informer cache yet
+	if synthd := comp.Status.CurrentSynthesis.Synthesized; synthd != nil {
+		delta := time.Since(synthd.Time)
+		if delta < time.Second*5 {
```

Comment:
Kinda scary that we rely on wall clock time for this, but I guess there's no better way. _Should_ be okay because the synth time is written by the Eno controller and should be (in most scenarios barring leader election) the same wall clock.

Reply:
Ah, it's also fine because there's a passthrough check below.

Reply:
This is just an optimization to avoid making the apiserver roundtrip that happens next :)